src/coreclr/src/System.Private.CoreLib/src/System/Threading/ProcessorIdCache.cs

```diff
@@ -0,0 +1,127 @@
+            int currentProcessorIdCache = t_currentProcessorIdCache--;
+            if ((currentProcessorIdCache & ProcessorIdCacheCountDownMask) == 0)
+            {
+                return RefreshCurrentProcessorId();
+            }
+
+            return currentProcessorIdCache >> ProcessorIdCacheShift;
+        }
+
+        // If GetCurrentProcessorNumber takes any nontrivial time (compared to TLS access), return false.
+        // Check more than once - to make sure it was not because TLS was delayed by GC or a context switch.
+        internal static bool ProcessorNumberSpeedCheck()
+        {
+            // NOTE: We do not check the frequency of the Stopwatch.
+            //       The frequency often does not match the actual timer refresh rate anyways.
+            //       If the resolution, precision or access time to the timer are inadequate for our measures here,
+            //       the test will fail anyways.
+
+            // default values pretend to be very long samples with default ratio
+            double minID = Stopwatch.Frequency * 50;  // 50 sec
+            double minTLS = Stopwatch.Frequency;      // 1 sec
+
+            // warm up the code paths.
+            UninlinedThreadStatic();
+            // also check if API is actually functional (-1 means not supported)
+            if (Thread.GetCurrentProcessorNumber() < 0)
+            {
+                s_processorIdRefreshRate = int.MaxValue;
+                return false;
+            }
+
+            long oneMicrosecond = Stopwatch.Frequency / 1000000;
+            for (int i = 0; i < 10; i++)
+            {
+                // we will measure at least 16 iterations and at least 1 microsecond
+                long t1;
+                int iters = 8;
+                do
+                {
+                    iters *= 2;
+                    t1 = Stopwatch.GetTimestamp();
+                    for (int j = 0; j < iters; j++)
+                    {
+                        Thread.GetCurrentProcessorNumber();
+                    }
+                    t1 = Stopwatch.GetTimestamp() - t1;
+                } while (t1 < oneMicrosecond);
+
+                minID = Math.Min(minID, (double)t1 / iters);
+
+                // we will measure at least 1 microsecond,
+                // and use at least 1/2 of ProcID iterations
+                // we assume that TLS can't be more than 2x slower than ProcID
+                iters /= 2;
+                do
+                {
+                    t1 = Stopwatch.GetTimestamp();
+                    for (int j = 0; j < iters; j++)
+                    {
+                        UninlinedThreadStatic();
+                    }
+                    t1 = Stopwatch.GetTimestamp() - t1;
+                    iters *= 2;
+                } while (t1 < oneMicrosecond) ;
+
+                minTLS = Math.Min(minTLS, (double)t1 / iters);
+            }
+
+            s_processorIdRefreshRate = Math.Min((int)(minID / minTLS), MaxIdRefreshRate);
```

Comment:
Do we need to crop it on the low-end as well to protect against bad clocks, etc.?

`Math.Clamp`

Reply:
I think both mins will be positive and not large (relative to the range of Double) values. If it goes below "fast" threshold, we will not be using cache, so effective refresh seems to be between "fast" threashold and MaxIdRefresh

---
In reply to: [357429978](https://github.com/dotnet/runtime/pull/467#discussion_r357429978) [](ancestors = 357429978)